keymanager/src/churp/handler.rs

```diff
@@ -0,0 +1,263 @@
+//! CHURP handler.
+use std::{
+    any::Any,
+    num::NonZeroUsize,
+    sync::{Arc, RwLock},
+};
+
+use anyhow::Result;
+use lru::LruCache;
+use p256::elliptic_curve::{group::GroupEncoding, Group};
+use rand::rngs::OsRng;
+use sp800_185::KMac;
+
+use oasis_core_runtime::{
+    self,
+    common::{
+        crypto::signature::{PublicKey, Signer},
+        namespace::Namespace,
+    },
+    consensus::{
+        beacon::EpochTime, state::keymanager::churp::GroupID,
+        verifier::Verifier as ConsensusVerifier,
+    },
+    identity::Identity as EnclaveIdentity,
+    storage::KeyValue,
+};
+
+use secret_sharing::{
+    churp::{Dealer, DealerParams, NistP384},
+    vss::matrix::VerificationMatrix,
+};
+
+use crate::beacon::State as BeaconState;
+
+use super::{
+    storage::Storage, ApplicationRequest, Error, InitRequest, SignedApplicationRequest,
+    State as ChurpState,
+};
+
+/// A handoff interval that disables handoffs.
+pub const HANDOFFS_DISABLED: EpochTime = 0xffffffffffffffff;
+
+/// Context used for the application request signature.
+const APPLICATION_REQUEST_SIGNATURE_CONTEXT: &[u8] =
+    b"oasis-core/keymanager/churp: application request";
+
+const CHECKSUM_VERIFICATION_MATRIX_CUSTOM: &[u8] =
+    b"oasis-core/keymanager/churp: verification matrix";
+
+/// The maximum number of CHURP dealers kept in the cache.
+const DEALERS_CACHE_SIZE: usize = 10;
+
+/// Key manager application that implements churn-robust proactive secret
+/// sharing scheme (CHURP).
+pub struct Churp {
+    runtime_id: Namespace,
+    node_id: PublicKey,
+
+    // Runtime attestation key signer.
+    signer: Arc<dyn Signer>,
+    // Storage handler.
+    storage: Storage,
+
+    churp_verifier: ChurpState,
+    beacon_verifier: BeaconState,
+
+    dealers: RwLock<LruCache<(Namespace, u64), Arc<dyn Any + Send + Sync>>>,
+}
+
+impl Churp {
+    pub fn new(
+        runtime_id: Namespace,
+        identity: Arc<EnclaveIdentity>,
+        consensus_verifier: Arc<dyn ConsensusVerifier>,
+        storage: Arc<dyn KeyValue>,
+    ) -> Self {
+        let storage = Storage::new(storage);
+        let signer: Arc<dyn Signer> = identity.clone();
+        let churp_verifier = ChurpState::new(consensus_verifier.clone());
+        let beacon_verifier = BeaconState::new(consensus_verifier.clone());
+
+        let dealers = RwLock::new(LruCache::new(
+            NonZeroUsize::new(DEALERS_CACHE_SIZE).unwrap(),
+        ));
+
+        Self {
+            signer,
+            runtime_id,
+            node_id: Default::default(),
```

Comment:
Will this be filled in later? It can be fetched from the host.

Reply:
Yes, that was the plan (but I forgot about it). Will add another commit.